docs/developers/debugging/compile_time_regressions.md

```diff
@@ -0,0 +1,202 @@
+  Identifying the scale of the regression can help set the priority to
+  investigate it.
+
+* **What is the full compile command?**
+
+  Try to extract the input program and full list of flags passed to the
+  compiler binary so that others can reproduce what you're seeing. Try to
+  distill this as much as possible to using just native tools (no Python or
+  other framework layers).
+
+* **What environment is the compiler running in?**
+
+  Are you using a `Debug` build, or a release build? What operating system and
+  size machine is running the compiler (e.g. Linux developer machine, or a
+  smaller system)?
+
+## Culprit finding and bisecting
+
+If you only have a rough idea of when something changed and want to narrow that
+down to a specific code change, bisecting can help.
+
+### Running `git bisect`
+
+Building the compiler from source and using
+[`git bisect`](https://git-scm.com/docs/git-bisect) will let you pinpoint
+specific commits in IREE, though it typically won't let you step through changes
+in submodules (e.g. MLIR updates in `third_party/llvm-project/`).
+
+__Tip__: [Configure ccache](../developing_iree/ccache.md) if you'll be rebuilding the compiler while bisecting
+
+A manual workflow with `git bisect` looks like this:
+
+```bash
+git bisect start --first-parent
+git bisect good [<rev>]
+git bisect bad [<rev>]
+
+# Read the prompts from the command as it runs
+# At each step, test the compiler:
+#   git submodule update
+#   cmake --build build/ --target iree-compile
+#   ./build/tools/iree-compile <args>
+#       attach Tracy, observe timing, print IR, etc. to determine if fast or slow
+#       if fast, `git bisect good`
+#       if slow, `git bisect bad`
+#   repeat
+```
+
+An automated workflow can use `git bisect run` and a script:
+
+```shell
+# run_bisect.sh
+git submodule update
+cmake --build build/ --target iree-compile
+# Other logic here
+```
+
+```bash
+git bisect start --first-parent
+git bisect good [<rev>]
+git bisect bad [<rev>]
+git bisect run run_bisect.sh
+```
+
+Other sample scripts:
+
+<details><summary>Compile executable sources individually with a timeout:</summary>
+
+```bash
+#!/bin/bash
+
+set -e
+set -o pipefail
+
+# Configure this based on what your thresholds are.
+TIMEOUT_SECONDS_FOR_COMPILING_EACH_SOURCE=10
+
+# Store git version hash, so we can dump artifacts to unique directories later.
+GIT_SHA="$(git rev-parse --short HEAD)"
+
+echo "** Building iree-compile at ${GIT_SHA} **"
+
+# The `git bisect` command only checks out a commit, so update submodules.
+git submodule update
+
+# Build the compiler. You'll want ccache configured to make this fast!
+cmake --build ../iree-build/ --target iree-compile
+
+# Run the compiler, dumping executable sources and stopping.
+SOURCES_DIR="../iree-tmp/sources-${GIT_SHA}"
+echo "** Running iree-compile at ${GIT_SHA}, dumping sources to ${SOURCES_DIR} **"
+../iree-build/tools/iree-compile \
+    ~/code/iree-tmp/t5_large_imported_2023_04_20.mlirbc \
```

Comment:
Could we make this variable too?

Reply:
Perhaps also skip if compile fails here already

Reply:
Yep, input file could be a variable... but then I'd also want to make `--iree-hal-target-backends=`, `--iree-input-type=`, and a few other flags variables... I think people will just need to fork the script and put their own compile flags in it.

Reply:
Added some variables in a section that is easy to edit.